### PR TITLE
Feature/07 2024 add scheduler checkin

### PR DESCRIPTION
**Pull request by solomoreal**
## Add API Check-in-Checkout Feature For Scheduler
- add check-in and checkout time to the scheduler model
- create checking/checkout routes and controller Methods
- write test

### CODE REVIEW

File: app/Http/API/V1/Controllers/SchedulerController.php [hunk 1]
```diff
@@ -28,4 +28,28 @@ public function show(Scheduler $scheduler): JsonResponse
...
+        }
+
+        $scheduler->checkedin_at = now();
+        $scheduler->update();
+
+        return $this->success('Your check in was successful');
+    }
+
+    public function checkOut(Scheduler $scheduler): JsonResponse
+    {
```
[the remark this answers is not in the document]
Comment (worldwayii) [reply app/Http/API/V1/Controllers/SchedulerController.php:45]: Then in the method, we $this->authorize(auth()->user(), $scheduler)

File: app/Models/Scheduler.php [hunk 2]
```diff
@@ -111,5 +115,12 @@ public function getUnpublishedSchedulersForEmployee(Employee $employee){
         return $result;
     }
 
+    public function checkInTime(){
```
Comment (worldwayii) [inline app/Models/Scheduler.php:118]: This is very superficial but improves readability please can we always have the open braces in the next line? 
`public function checkInTime()`
`{`
`}`

File: routes/api.php [hunk 3]
```diff
@@ -37,4 +37,9 @@
 
     Route::get('scheduler/me', [SchedulerController::class, 'index'])->name('schedules');
     Route::get('scheduler/me/{scheduler}', [SchedulerController::class, 'show'])->name('schedule');
+
+    Route::controller(UserController::class)->prefix('scheduler/me')->group(function () {
+        Route::put('checkin/{scheduler}', 'checkIn')->name('scheduler.checkin');
```
Comment (worldwayii) [inline routes/api.php:42]: can the name be `schedule.checkin` & `scheduler.checkout` (singular)?

**Review by worldwayii — CHANGES_REQUESTED** [review]

File: app/Http/API/V1/Controllers/SchedulerController.php [hunk 4]
```diff
@@ -28,4 +28,28 @@ public function show(Scheduler $scheduler): JsonResponse
...
+        if($scheduler->checkOutTime() != null){
+            return $this->wrongRequestType('You have already checked out for this shift');
+        }
+
```
[the remark this answers is not in the document]
Comment (solomoreal) [reply app/Http/API/V1/Controllers/SchedulerController.php:49]: noted, I will go with option 2 for now, we can update it when the necessary changes are made to the UI

File: app/Models/Scheduler.php (hunk 2 above)
Comment (solomoreal) [reply app/Models/Scheduler.php:118]: noted

File: app/Http/API/V1/Controllers/SchedulerController.php [hunk 5]
```diff
@@ -28,4 +28,28 @@ public function show(Scheduler $scheduler): JsonResponse
                 $scheduler->load('employee', 'company')
             ], 'Schedule fetched successfully.');
     }
+
+    public function checkIn(Scheduler $scheduler): JsonResponse
+    {
```
[the remark this answers is not in the document]
Comment (solomoreal) [reply app/Http/API/V1/Controllers/SchedulerController.php:33]: noted

File: app/Http/API/V1/Controllers/SchedulerController.php [hunk 6]
```diff
@@ -28,4 +30,98 @@ public function show(Scheduler $scheduler): JsonResponse
...
+            return $this->internalServerError('Clockin Failed Please Try Again!.', $e->getMessage());
+        }
+
+    }
+
+    public function checkOut(Scheduler $scheduler): JsonResponse
+    {
+        //$this->authorize('update', $scheduler);
+
+        if (request()->user()->cannot('update', $scheduler)) {
```
Comment (worldwayii) [inline app/Http/API/V1/Controllers/SchedulerController.php:62]: can you use `$this->authorize('update', $scheduler)` ( checkOut/ checkIn) in you  policy checks rather?
it's a shorter syntax and works fine. You can add test coverage for wrong user to see if the policy check catches them?

File: app/Http/API/V1/Controllers/SchedulerController.php [hunk 7]
```diff
@@ -28,4 +30,98 @@ public function show(Scheduler $scheduler): JsonResponse
...
+            if(now()->lt($scheduler->end_at)){
+                return [
+                    'status' => false,
+                    'message' => 'You can not clock out before your shift ends',
+                ];
+            }
+        }
+
+
+        return [
```
Comment (worldwayii) [inline app/Http/API/V1/Controllers/SchedulerController.php:118]: what is the need for this return with empty string `' '` message? Just return `true`?

File: tests/Integration/Http/API/V1/Controllers/SchedulerControllerTest.php [hunk 8]
```diff
@@ -126,4 +127,139 @@ public function testIndexWhenUnauthenticatedDoesNotEmployeeSchedules()
...
+        $response = $this->withHeaders([
+            'Authorization' => 'Bearer ' . $token,
+        ])->putJson(route('schedule.checkout', $scheduler->id));
+
+        $response->assertStatus(405)
+            ->assertJson([
+                'message' => 'You can not clock out before your shift ends',
+            ]);
+    }
+
```
Comment (worldwayii) [inline tests/Integration/Http/API/V1/Controllers/SchedulerControllerTest.php:264]: NIce test coverage, can we have one or 2 to cover policy check? ` testCheckOutFailsIfWhenUserIsWorng` && `` testCheckInFailsIfWhenUserIsWorng``

**Review by worldwayii — APPROVED** [review]

**Pull request merged**